File: plugins/grok/scripts/lib/ref-prep.mjs
/**
 * Getting an edit's source images to Grok with their detail.
 *
 * The Grok CLI sends image_edit a JPEG or PNG of up to 400 KB as it is, and
 * shrinks anything bigger — or in another format — to 768 px first
 * (`compress_reference` in its image_edit tool). A photo straight off a phone
 * would lose most of its detail. So such a reference is re-encoded here, by
 * `scripts/refprep.py` (Python with Pillow), at up to 1536 px in under 400 KB,
 * and Grok gets that copy instead. Without Python and Pillow the reference
 * goes as it is, and a note says the CLI shrinks it.
 */

import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

import { findPython } from "./chroma.mjs";
import { runProgram } from "./ffmpeg.mjs";

const SCRIPT = fileURLToPath(new URL("../refprep.py", import.meta.url));

/** The most the Grok CLI sends to image_edit as it is. */
export const PASS_THROUGH_BYTES = 400 * 1024;
/** The longest side a prepared reference keeps; a live edit on 2026-09-25 sent a 1536×864 copy and it was taken. */
export const PREPARED_MAX_SIDE = 1536;
/** What the Grok CLI shrinks a reference to when it has to do it itself. */
const CLI_REFERENCE_SIDE = 768;

const MISSING_LIBRARIES = 3;

/** A JPEG or PNG, going by its first bytes. */
function isJpegOrPng(file) {
  let head;
  try {
    const handle = fs.openSync(file, "r");
    try {
      head = Buffer.alloc(8);
      fs.readSync(handle, head, 0, 8, 0);
    } finally {
      fs.closeSync(handle);
    }
  } catch {
    return false;
  }
  return (head[0] === 0xff && head[1] === 0xd8) || head.readUInt32BE(0) === 0x89504e47;
}

/** Whether the Grok CLI would send `image` (a path or a data: URL) to image_edit untouched. */
export function passesThrough(image) {
  if (image.startsWith("data:")) {
    return true; // left to the CLI, which handles data: URLs itself
  }
  return fs.statSync(image).size <= PASS_THROUGH_BYTES && isJpegOrPng(image);
}

/**
 * Prepare each of `images` that the Grok CLI would shrink, writing the copies
 * into `workDir` (which the caller removes after the run).
 *
 * Returns `{ images, prepared, notes }`: the paths to send, in order; what was
 * prepared, for the manifest (`{ source, width, height, bytes, format }`); and
 * notes for the result.
 */
export async function prepareEditReferences(images, { workDir }) {
  const sent = [];
  const prepared = [];
  const notes = [];
  let python;

  for (const [index, image] of images.entries()) {
    if (passesThrough(image)) {
      sent.push(image);
      continue;
    }
    const name = path.basename(image);
    const shrunk = `the Grok CLI shrinks it to ${CLI_REFERENCE_SIDE} px`;
    try {
      python ??= findPython();
    } catch {
      notes.push(`Note: ${name} is over 400 KB or not a JPEG/PNG, and without Python the plugin cannot prepare it, so ${shrunk}.`);
      sent.push(image);
      continue;
    }

    fs.mkdirSync(workDir, { recursive: true });
    const output = path.join(workDir, `ref-${index + 1}-${path.parse(image).name}.prepared`);
    let code;
    let stdout;
    let stderr;
    try {
      ({ code, stdout, stderr } = await runProgram(python, [SCRIPT, image, "--output", output, "--max-side", String(PREPARED_MAX_SIDE)]));
    } catch (error) {
      notes.push(`Note: the plugin could not prepare ${name} (${error.message}), so ${shrunk}.`);
      sent.push(image);
      continue;
    }
    if (code === 0) {
      const result = JSON.parse(stdout);
      const file = `${output}.${result.format === "png" ? "png" : "jpg"}`;
      fs.renameSync(output, file);
      sent.push(file);
      prepared.push({ source: image, width: result.width, height: result.height, bytes: result.bytes, format: result.format });
      notes.push(
        `Note: ${name} went to Grok as a ${result.width}×${result.height} ${result.format.toUpperCase()} of ${Math.round(result.bytes / 1024)} KB, ` +
          `prepared by the plugin; as it was, ${shrunk}.`
      );
      continue;
    }
    const reason = code === MISSING_LIBRARIES ? `${python} lacks Pillow` : stderr.trim().split("\n").at(-1) || `refprep.py exit ${code}`;
    notes.push(`Note: the plugin could not prepare ${name} (${reason}), so ${shrunk}.`);
    sent.push(image);
  }

  return { images: sent, prepared, notes };
}

File: plugins/grok/scripts/refprep.py
#!/usr/bin/env python3
"""Fit an edit reference into what the Grok CLI sends to image_edit as it is.

    refprep.py INPUT --output PATH [--max-bytes N] [--max-side N] [--min-side N]

The Grok CLI passes a JPEG or PNG of up to 400 KB to image_edit untouched and
shrinks anything else to 768 px. This writes INPUT upright (EXIF orientation
applied) at the largest size, up to --max-side, that fits in --max-bytes: a
JPEG, or a PNG when the image has transparency. It never enlarges. Prints one
JSON object: the size and format written, and the source's size.

Exit codes: 0 done, 1 refused (it does not fit even at --min-side; the reason
is on stderr), 2 failed (a traceback on stderr), 3 Pillow is missing (JSON
{"missing": [...]} on stderr).
"""

import argparse
import io
import json
import sys
import traceback

REFUSED = 1
FAILED = 2
MISSING_LIBRARIES = 3

# Tried from the top down: the first side that fits in the byte budget wins.
SIDES = (1536, 1280, 1024, 768)
JPEG_QUALITIES = (90, 85, 80, 75, 70, 65, 60)


def has_alpha(image):
    return image.mode in ("RGBA", "LA", "PA") or (image.mode == "P" and "transparency" in image.info)


def fitted(image, side, Image):
    width, height = image.size
    scale = side / max(width, height)
    if scale >= 1:
        return image
    return image.resize((max(1, round(width * scale)), max(1, round(height * scale))), Image.Resampling.LANCZOS)


def encode(image, alpha):
    """Yield (bytes, format) from the best quality down."""
    if alpha:
        buffer = io.BytesIO()
        image.convert("RGBA").save(buffer, "PNG", optimize=True)
        yield buffer.getvalue(), "png"
        return
    rgb = image.convert("RGB")
    for quality in JPEG_QUALITIES:
        buffer = io.BytesIO()
        rgb.save(buffer, "JPEG", quality=quality, optimize=True)
        yield buffer.getvalue(), "jpeg"


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("input")
    parser.add_argument("--output", required=True)
    parser.add_argument("--max-bytes", type=int, default=400 * 1024)
    parser.add_argument("--max-side", type=int, default=SIDES[0])
    parser.add_argument("--min-side", type=int, default=SIDES[-1])
    args = parser.parse_args()

    try:
        from PIL import Image, ImageOps
    except ImportError:
        sys.stderr.write(json.dumps({"missing": ["Pillow"]}))
        return MISSING_LIBRARIES

    with Image.open(args.input) as source:
        image = ImageOps.exif_transpose(source)
        image.load()
    alpha = has_alpha(image)
    longest = max(image.size)
    sides = sorted({min(side, longest) for side in (args.max_side, *SIDES) if args.min_side <= side <= args.max_side}, reverse=True)

    for side in sides:
        candidate = fitted(image, side, Image)
        for data, kind in encode(candidate, alpha):
            if len(data) <= args.max_bytes:
                with open(args.output, "wb") as out:
                    out.write(data)
                json.dump(
                    {
                        "width": candidate.size[0],
                        "height": candidate.size[1],
                        "bytes": len(data),
                        "format": kind,
                        "sourceWidth": image.size[0],
                        "sourceHeight": image.size[1],
                    },
                    sys.stdout,
                )
                return 0

    sys.stderr.write(f"{args.input} does not fit in {args.max_bytes} bytes even at {sides[-1] if sides else longest} px.\n")
    return REFUSED


if __name__ == "__main__":
    try:
        sys.exit(main())
    except SystemExit:
        raise
    except Exception:  # noqa: BLE001 — any failure is reported, not hidden
        traceback.print_exc()
        sys.exit(FAILED)
